docs/proposals/server-side-apply-field-managers.md

```diff
@@ -0,0 +1,123 @@
+
+- Manage only specific fields they care about on a shared resource
+- Avoid deleting or overwriting fields that are managed by other Applications
+
+## Motivation
+
+There exist situations where disparate Applications need to add or remove configuration from a shared Kubernetes resource. Server-side apply supports this behavior when different field managers are used.
+
+## Goals
+
+All following goals should be achieve in order to conclude this proposal:
+
+#### [G-1] Applications can apply changes to a shared resource without disturbing existing fields
+
+A common use case of server-side apply is the ability to manage only particular fields on a share Kubernetes resource, while leaving everything else in tact. This requires a unique field manager for each identity that shares the resource.
+
+#### [G-2] Applications that are destroyed only remove the fields they manage from shared resources
+
+A delete operation should undo only the additions or updates it made to a shared resource, unless that resource has `Prune=true`.
+
+#### [G-3] Users can define a field manager as a sync option
+
+Some users may rely on the current behavior emanating from the use of the same field manager across all Applications. The current behavior being that each server side apply sync overwrites all fields on a resource. In other words, "latest sync wins."
+
+## Non-Goals
+
+N/A
+
+## Proposal
+
+1. Add a new sync option named `FieldManager` that accepts a string up to 128 characters in length. This can only be set on an individual resource. Don't allow this sync option to be set on the Application: accidentally overriding the resource-level field manager may have undesirable side effects like leaving orphaned fields behind. When a sync is performed for a resource and server side-apply is enabled, it uses the `FieldManager` if it is set, otherwise it defaults to the hard-coded `argocd-controller` field manager.
+
+1. Like other sync options, add a corresponding text field in the ArgoCD UI to let users specify a field manager for a server-side apply sync. This text field should only be visible when a single resource is being synced.
+
+1. Change the removal behavior for shared resources. When a resource with a custom field manager is "removed", it instead removes only the fields managed by its field manager from the shared resource by sending an empty "fully specified intent" using server-side apply. You can fully delete a shared resource by setting `Prune=true` at the resource level.
+
+1. Add documentation suggesting that users might want to consider changing the permissions on the ArgoCD role to disallow the `delete` and `update` verbs on shared resources. Server-side apply will always use `patch`, and removing `delete` and `update` helps prevent users from errantly wiping out changes made from other Applications.
```

Comment:
Even if we do suggest that in our docs, we will still need to handle the transitions properly:
- standard manager to custom manager
- custom manager to standard manager

Reply:
Same point as above, I think immutability would solve this issue. Require the resource to be "deleted" in the application and recreated with a new field manager if you want to change it.

Reply:
Argo CD runs ServerSide Apply with `--force-conflicts` flag. This section in the Kubernetes doc doesn't apply in this case. It would be great to have this transition scenario added in the demo section as well as I am still not sure about the exact behaviour. We have:

1. Argo CD App1 syncs the main resource with ServerSide Apply with manager `argocd-controller` and `--force-conflict-flag`.
2. Argo CD App2 syncs different fields (`another-field`) in the same resource with another manager (`another-manager`)
3. Argo CD App2 switches to default argocd manager (`argocd-controller`), mutates the `another-field` and syncs with ServerSide Apply (at this point all fields in the resource are owned by the `argocd-controller`)
4. Argo CD App2 switches to again to `another-manager`, mutates the `another-field` and syncs with ServerSide Apply

At this stage I am not sure about:
- how the fields ownership will look like
- how App1 and App2 will behave as it seems that they will be in constant conflict with no easy way repair.

Reply:
Assuming that App2 _only_ has `another-field` in its fully specified intent, the end result after steps 1-4 would be just the fields from App2. Just `another-field`.

This is because in step 3, App2 would have wiped out the fields from App1 because it used the same field manager (`argocd-controller`) as App1 and didn't specify any fields that App1 applied. I'll get another demo for this.

Reply:
Assuming that App2 _only_ has `another-field` in its fully specified intent, the end result after steps 1-4 would be just the fields from App2. Just `another-field`.

This is because in step 3, App2 would have wiped out the fields from App1 because it used the same field manager (`argocd-controller`) as App1 and didn't specify any fields that App1 applied. I'll get another demo for this.

Reply:
I don't know if I should include this in the proposal as an example of "orphaned" fields. I think the real issue demonstrated here is that an Application could overwrite a field set by another application. But that's already the case without `FailOnSharedResource=true`. Giving users the ability to change the field manager allows them to avoid overwriting shared objects. And setting the field manager to match another application is essentially declaring that you share the identity.

We could take an opinionated stance and assert that each application should use a unique field manager, and remove the ability for users to set it manually. This way, there is no risk of orphaning fields. I think this lines up better with ArgoCD's ownership model. It _may_ break behavior for users that use different Applications to fully overwrite shared resources in a "last sync wins" manner. 

```bash
#!/usr/bin/env bash

# Create kind cluster
kind create cluster

# Create service in App1
cat <<EOF > /tmp/app1-service.yaml
apiVersion: v1
kind: Service
metadata:
  name: nginx-service
spec:
  selector:
    app: nginx
  ports:
    - protocol: TCP
      port: 80
      targetPort: 8080
EOF

# Create service in App2
cat <<EOF > /tmp/app2-service.yaml
apiVersion: v1
kind: Service
metadata:
  name: nginx-service
spec:
  ports:
    - protocol: TCP
      name: foo
      port: 1000
      targetPort: 8080
EOF

# Sync App1 with argocd-controller
kubectl --context kind-kind apply -f /tmp/app1-service.yaml --force-conflicts --server-side --field-manager argocd-controller

# Sync App2 with another-manager
kubectl --context kind-kind apply -f /tmp/app2-service.yaml --force-conflicts --server-side --field-manager another-manager

# Modify App2 service
cat <<EOF > /tmp/app2-service.yaml
apiVersion: v1
kind: Service
metadata:
  name: nginx-service
spec:
  ports:
    - protocol: TCP
      name: bar
      port: 2000
      targetPort: 8080
EOF

# Sync App2 with argocd-controller
kubectl --context kind-kind apply -f /tmp/app2-service.yaml --force-conflicts --server-side --field-manager argocd-controller

# Sync App2 with argocd-controller
kubectl --context kind-kind apply -f /tmp/app2-service.yaml --force-conflicts --server-side --field-manager argocd-controller

# Modify App2 service
cat <<EOF > /tmp/app2-service.yaml
apiVersion: v1
kind: Service
metadata:
  name: nginx-service
spec:
  type: LoadBalancer
  ports:
    - protocol: TCP
      name: buzz
      port: 3000
      targetPort: 8080
EOF

# Sync App2 with another-manager
kubectl --context kind-kind apply -f /tmp/app2-service.yaml --force-conflicts --server-side --field-manager another-manager

# Get the service. Notice that the selector is missing.
kubectl --context kind-kind get service nginx-service -oyaml

# Destroy kind cluster
kind delete cluster
```

Reply:
Great! Please add this test script at the bottom of your proposal with a note that once implemented, the Argo CD feature needs to have the same behaviour validated as part of its integration test suite.